pkg/pool-weighted/contracts/managed/ManagedPoolCircuitBreakerLib.sol

```diff
@@ -0,0 +1,248 @@
+        _require(params.upperBoundPercentage <= _MAX_BOUND_PERCENTAGE, Errors.INVALID_CIRCUIT_BREAKER_BOUNDS);
+        _require(
+            params.upperBoundPercentage == 0 || params.upperBoundPercentage >= params.lowerBoundPercentage,
+            Errors.INVALID_CIRCUIT_BREAKER_BOUNDS
+        );
+
+        // Set the basic parameters, and chain to `_setCircuitBreakerState` for the rest.
+        circuitBreakerState = circuitBreakerState
+            .insertUint(params.referenceBptPrice, _REFERENCE_BPT_PRICE_OFFSET, _BPT_PRICE_WIDTH)
+            .insertUint(
+            params.currentWeightFactor.compress(_WEIGHT_FACTOR_WIDTH),
+            _WEIGHT_FACTOR_OFFSET,
+            _WEIGHT_FACTOR_WIDTH
+        );
+
+        return _setCircuitBreakerState(circuitBreakerState, params);
+    }
+
+    // This function is only needed to address stack-too-deep issues
+    function _setCircuitBreakerState(bytes32 circuitBreakerState, CircuitBreakerParams memory params)
+        private
+        pure
+        returns (bytes32)
+    {
+        // Add the raw percentage boundaries
+        circuitBreakerState = circuitBreakerState
+            .insertUint(
+            params.lowerBoundPercentage.compress(_BOUND_PERCENTAGE_WIDTH, _MAX_BOUND_PERCENTAGE),
+            _LOWER_BOUND_PCT_OFFSET,
+            _BOUND_PERCENTAGE_WIDTH
+        )
+            .insertUint(
+            params.upperBoundPercentage.compress(_BOUND_PERCENTAGE_WIDTH, _MAX_BOUND_PERCENTAGE),
+            _UPPER_BOUND_PCT_OFFSET,
+            _BOUND_PERCENTAGE_WIDTH
+        );
+
+        // Precompute and store the conversion ratios; if the weights aren't changing, we can use these directly.
+        (uint256 lowerBoundRatioCache, uint256 upperBoundRatioCache) = _getBoundaryConversionRatios(
+            params.lowerBoundPercentage,
+            params.upperBoundPercentage,
+            params.currentWeightFactor
+        );
+
+        return
+            circuitBreakerState
+                .insertUint(
+                lowerBoundRatioCache.compress(_BOUND_RATIO_CACHE_WIDTH, _MAX_BOUND_PERCENTAGE),
+                _LOWER_BOUND_RATIO_CACHE_OFFSET,
+                _BOUND_RATIO_CACHE_WIDTH
+            )
+                .insertUint(
+                upperBoundRatioCache.compress(_BOUND_RATIO_CACHE_WIDTH, _MAX_BOUND_PERCENTAGE),
+                _UPPER_BOUND_RATIO_CACHE_OFFSET,
+                _BOUND_RATIO_CACHE_WIDTH
+            );
+    }
+
+    // Convert percentage bounds to BPT price bounds
+    function _getBoundaryConversionRatios(
+        uint256 lowerBoundPercentage,
+        uint256 upperBoundPercentage,
+        uint256 currentWeightFactor
+    ) private pure returns (uint256 lowerBoundRatioCache, uint256 upperBoundRatioCache) {
+        lowerBoundRatioCache = LogExpMath.pow(lowerBoundPercentage, currentWeightFactor);
+        upperBoundRatioCache = LogExpMath.pow(upperBoundPercentage, currentWeightFactor);
+    }
```

Comment:
It's a function because it's called in two places (when setting and getting). To make it more lenient (i.e., fail less often), I can use powDown for the lower bound and powUp for the upper bound.